### PR TITLE
update transport protocol to use TLS

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: protocol/simplex-messaging.md [hunk 1]
```diff
@@ -84,12 +84,15 @@ The [ABNF][8] syntax of the message is:
 
 ```abnf
 queueInfo = %s"smp::" smpServer "::" queueId "::" encryptionKey
-smpServer = srvHost [":" port] ["#" serverKeyHash]
+smpServer = srvHost [":" port] "#" serverIdentity
 srvHost = <hostname> ; RFC1123, RFC5891
 port = 1*DIGIT
 serverKeyHash = encoded
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:90]:
```suggestion
serverIdentity = encoded
```

File: protocol/simplex-messaging.md [hunk 2]
```diff
@@ -98,7 +101,7 @@ encoded = <base64 encoded>
 
 `port` is optional, the default TCP port for SMP protocol is 5223.
 
-`serverKeyHash` is an optional hash of the server transport key used during transport handshake (see [Appendix A](#appendix-a)).
+`serverIdentity` is a required hash of the server certificate SPKI (without line breaks) used by the client to validate server certificate during transport handshake (see [Appendix A](#appendix-a)).
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:104]: SPKI is "Simple public key infrastructure"? I understand what the sentence means, but not sure how to read it.. Should it be "SPKI certificate"?

File: protocol/simplex-messaging.md [hunk 3]
```diff
@@ -348,16 +351,18 @@ Each transmission between the client and the server must have this format/syntax
 
 ```abnf
 transmission = [signature] SP signed SP pad ; pad to the fixed block size
-signed = [corrId] SP [queueId] SP cmd
+signed = sessionIdentifier SP [corrId] SP [queueId] SP cmd ; corrId is required in client commands and server responses,
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:354]: Define `sessionIdentifier` here? It is way below

File: protocol/simplex-messaging.md [hunk 4]
```diff
@@ -639,52 +645,45 @@ ok = %s"OK"
 
 Both the recipient and the sender can use TCP or some other, possibly higher level, transport protocol to communicate with the server. The default TCP port for SMP server is 5223.
 
-By default, the client and server should use the protocol presented below, that does not depend on a centralized certificate authority.
+For scenarios when meta-data privacy is critical, it is recommended that:
+- clients and servers communicating over Tor network,
+- establish a separate connection for each SMP queue,
+- send noise traffic (using PING command).
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:651]:
```suggestion
For scenarios when meta-data privacy is critical, it is recommended that clients and servers:
- communicate over Tor network,
- establish a separate connection for each SMP queue,
- send noise traffic (using PING command).
```

File: protocol/simplex-messaging.md (hunk 3 above)
Comment (efim-poberezkin) [reply protocol/simplex-messaging.md:354]: or add "See `sessionIdentifier` in [Appendix A - SMP transport protocol](#appendix-a)

File: protocol/simplex-messaging.md [hunk 5]
```diff
@@ -639,52 +645,45 @@ ok = %s"OK"
...
 
-Transport is encrypted with [AEAD-GCM][12] protocol with two random symmetric AES 256-bit keys and two random base IVs that will be agreed during the handshake. Both client and the server should maintain two 32-bit word counters, one for the sent and one for the received messages. The IV for each message should be computed by xor-ing the sequential message counter, starting from 0, with the first 32 bits of agreed base IV (the number is encoded in network byte order).
+The transport protocol should provide the following:
+- server authentication (by matching server certificate hash with `serverIdentity`),
+- forward secrecy (by encrypting the traffic using ephemeral keys agreed during transport handshake),
+- integrity (preventing data modification by the attacker without detection),
+- unique channel binding to include in the signed part of SMP transmissions.
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:657]:
```suggestion
- unique channel binding (`sessionIdentifier`) to include in the signed part of SMP transmissions.
```
?

File: protocol/simplex-messaging.md [hunk 6]
```diff
@@ -639,52 +645,45 @@ ok = %s"OK"
...
 
-To establish the session keys and base IVs, the server should have an asymmetric key pair generated during server deployment and unknown to the clients. The users should know the key hash (256 bits) in advance in order to be able to validate the server public key during transport connection handshake.
+By default, the client and server communicate using [TLS 1.3 protocol][13] restricted to:
+- TLS_AES_256_GCM_SHA384 cypher suite,
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:660]: why this SHA specifically?

File: protocol/simplex-messaging.md [hunk 7]
```diff
@@ -639,52 +645,45 @@ ok = %s"OK"
...
+- ed25519 and ed448 EdDSA algorithms for signatures,
+- x25519 and x448 ECDHE groups for key exchange.
+- servers must send only one self-signed certificate in the handshake, clients must abort the connection in case more than one certificate is sent.
+- server and client TLS configuration should not allow resuming the sessions.
 
-The handshake sequence is the following:
+During TLS handshake the client must validate that the hash of the server certificate SPKI block is equal to the `serverIdentity` the client received as part of SMP server address; if the server identity does not match the client must abort the connection.
```
Comment (efim-poberezkin) [inline protocol/simplex-messaging.md:666]: Also "SPKI certificate"?

**Review by efim-poberezkin — APPROVED** [review]

File: protocol/simplex-messaging.md (hunk 2 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:104]: I mean here SPKI block, that is not including line breaks and header/footer

File: protocol/simplex-messaging.md (hunk 3 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:354]: yes, the latter

File: protocol/simplex-messaging.md (hunk 4 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:651]: it's not exactly right... the client does all that, the server can only be deployed as Tor service though, right?...

File: protocol/simplex-messaging.md (hunk 7 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:666]: I think address is fine. See above about SPKI...

Comment (epoberezkin) [reply protocol/simplex-messaging.md:666]: I think address is fine. See above about SPKI...

File: protocol/simplex-messaging.md (hunk 3 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:354]: it's actually defined above

File: protocol/simplex-messaging.md (hunk 6 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:660]: I don't think there is suite with bigger SHA in TLS.

File: protocol/simplex-messaging.md (hunk 2 above)
Comment (epoberezkin) [reply protocol/simplex-messaging.md:104]: updated

**Pull request merged**